### PR TITLE
OkHttp 3.1.2, Retrofit 2.0.0-beta4, Stetho 1.3.0

### LINKED ISSUE #70: Http logging interceptor should be compiled only into debug build

### LINKED ISSUE #87: Retrofit 2 beta4 + OkHttp 3
^

### PR DESCRIPTION
**Pull request by artem-zinnatullin**
Closes #87, closes #70.

Required quite a lot changes because both OkHttp and Retrofit now in new packages w/ version number + OkHttp is now immutable but at the same time we have 2 interceptors for debug only build, so, yeah.

@vanniktech PTAL :)

### CODE REVIEW

**Comment by artem-zinnatullin** [comment]
Hm, looks like Robolectric has some problems with OkHttp 3.1.0

```
com.artemzin.qualitymatters.ui.fragments.ItemsFragmentTest > onTryAgainButtonClick_shouldRequestDataFromPresenter STANDARD_ERROR
    java.lang.IllegalStateException: Unable to extract the trust manager on okhttp3.internal.Platform$Android@483cacc0, sslSocketFactory is class sun.security.ssl.SSLSocketFactoryImpl
    	at okhttp3.OkHttpClient.<init>(OkHttpClient.java:187)
    	at okhttp3.OkHttpClient.<init>(OkHttpClient.java:60)
    	at okhttp3.OkHttpClient$Builder.build(OkHttpClient.java:719)
    	at com.artemzin.qualitymatters.network.NetworkModule.provideOkHttpClient(NetworkModule.java:34)
    	at com.artemzin.qualitymatters.network.NetworkModule_ProvideOkHttpClientFactory.get(NetworkModule_ProvideOkHttpClientFactory.java:27)
    	at com.artemzin.qualitymatters.network.NetworkModule_ProvideOkHttpClientFactory.get(NetworkModule_ProvideOkHttpClientFactory.java:10)
    	at dagger.internal.ScopedProvider.get(ScopedProvider.java:46)
    	at com.artemzin.qualitymatters.developer_settings.DeveloperSettingsModule_ProvideDeveloperSettingsModelImplFactory.get(DeveloperSettingsModule_ProvideDeveloperSettingsModelImplFactory.java:40)
    	at com.artemzin.qualitymatters.developer_settings.DeveloperSettingsModule_ProvideDeveloperSettingsModelImplFactory.get(DeveloperSettingsModule_ProvideDeveloperSettingsModelImplFactory.java:11)
    	at dagger.internal.ScopedProvider.get(ScopedProvider.java:46)
    	at com.artemzin.qualitymatters.developer_settings.DeveloperSettingsModule_ProvideDeveloperSettingsModelFactory.get(DeveloperSettingsModule_ProvideDeveloperSettingsModelFactory.java:21)
    	at com.artemzin.qualitymatters.developer_settings.DeveloperSettingsModule_ProvideDeveloperSettingsModelFactory.get(DeveloperSettingsModule_ProvideDeveloperSettingsModelFactory.java:7)
    	at dagger.internal.DoubleCheckLazy.get(DoubleCheckLazy.java:49)
    	at com.artemzin.qualitymatters.QualityMattersApp.onCreate(QualityMattersApp.java:51)
    	at org.robolectric.internal.ParallelUniverse.setUpApplicationState(ParallelUniverse.java:140)
    	at org.robolectric.RobolectricTestRunner.setUpApplicationState(RobolectricTestRunner.java:433)
    	at org.robolectric.RobolectricTestRunner$2.evaluate(RobolectricTestRunner.java:240)
    	at org.robolectric.RobolectricTestRunner.runChild(RobolectricTestRunner.java:188)
    	at org.robolectric.RobolectricTestRunner.runChild(RobolectricTestRunner.java:54)
    	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
    	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
    	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
    	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
    	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
    	at org.robolectric.RobolectricTestRunner$1.evaluate(RobolectricTestRunner.java:152)
    	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
    	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecuter.runTestClass(JUnitTestClassExecuter.java:105)
    	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecuter.execute(JUnitTestClassExecuter.java:56)
    	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassProcessor.processTestClass(JUnitTestClassProcessor.java:64)
    	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:50)
    	at sun.reflect.GeneratedMethodAccessor44.invoke(Unknown Source)
    	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    	at java.lang.reflect.Method.invoke(Method.java:497)
    	at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
    	at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
    	at org.gradle.messaging.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:32)
    	at org.gradle.messaging.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:93)
    	at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
    	at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:106)
    	at sun.reflect.GeneratedMethodAccessor43.invoke(Unknown Source)
    	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    	at java.lang.reflect.Method.invoke(Method.java:497)
    	at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
    	at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
    	at org.gradle.messaging.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:360)
    	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:54)
    	at org.gradle.internal.concurrent.StoppableExecutorImpl$1.run(StoppableExecutorImpl.java:40)
    	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    	at java.lang.Thread.run(Thread.java:745)

com.artemzin.qualitymatters.ui.fragments.ItemsFragmentTest > onTryAgainButtonClick_shouldRequestDataFromPresenter FAILED
    java.lang.RuntimeException: java.lang.IllegalStateException: Unable to extract the trust manager on okhttp3.internal.Platform$Android@483cacc0, sslSocketFactory is class sun.security.ssl.SSLSocketFactoryImpl

        Caused by:
        java.lang.IllegalStateException: Unable to extract the trust manager on okhttp3.internal.Platform$Android@483cacc0, sslSocketFactory is class sun.security.ssl.SSLSocketFactoryImpl
```

Will investigate later.

File: app/src/main/java/com/artemzin/qualitymatters/api/ApiModule.java [hunk 1]
```diff
@@ -31,17 +31,13 @@ public ChangeableBaseUrl provideChangeableBaseUrl() {
...
-                .addCallAdapterFactory(RxJavaCallAdapterFactory.create());
-
-        // Fail early: check Retrofit configuration at creation time
-        if (BuildConfig.DEBUG) {
-            builder.validateEagerly();
-        }
-
-        return builder.build().create(QualityMattersRestApi.class);
+                .addCallAdapterFactory(RxJavaCallAdapterFactory.create())
+                .validateEagerly(BuildConfig.DEBUG)  // Fail early: check Retrofit configuration at creation time in Debug build.
```
Comment (adelnizamutdinov) [inline app/src/main/java/com/artemzin/qualitymatters/api/ApiModule.java]: what about doing this in a unit test instead?

Comment (artem-zinnatullin) [inline app/src/main/java/com/artemzin/qualitymatters/api/ApiModule.java]: Good idea! The only downside of this is that it'll require providing `Retrofit.Builder`, but seems like worth it

File: app/src/debug/java/com/artemzin/qualitymatters/network/OkHttpInterceptorsModule.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    public HttpLoggingInterceptor provideHttpLoggingInterceptor() {
+        return new HttpLoggingInterceptor(message -> Timber.d(message));
+    }
+
+    @Provides @Named(LIST_OF_OK_HTTP_INTERCEPTORS) @Singleton @NonNull
+    public List<Interceptor> provideOkHttpInterceptors(@NonNull HttpLoggingInterceptor httpLoggingInterceptor) {
+        return singletonList(httpLoggingInterceptor);
+    }
+
+    @Provides @Named(LIST_OF_OK_HTTP_NETWORK_INTERCEPTORS) @Singleton @NonNull
```
Comment (vanniktech) [inline app/src/debug/java/com/artemzin/qualitymatters/network/OkHttpInterceptorsModule.java]: Why not simply create a custom qualifier annotation and use it here?
 
```java
@Qualifier
@Documented
@Retention(RUNTIME)
public @interface OkHttpNetworkInterceptors {
}
```

File: app/src/debug/java/com/artemzin/qualitymatters/network/OkHttpInterceptorsModule.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+@Module
+public class OkHttpInterceptorsModule {
+
+    // Provided as separate dependency for Developer Settings to be able to change HTTP log level at runtime.
+    @Provides @Singleton @NonNull
+    public HttpLoggingInterceptor provideHttpLoggingInterceptor() {
+        return new HttpLoggingInterceptor(message -> Timber.d(message));
+    }
+
+    @Provides @Named(LIST_OF_OK_HTTP_INTERCEPTORS) @Singleton @NonNull
```
Comment (vanniktech) [inline app/src/debug/java/com/artemzin/qualitymatters/network/OkHttpInterceptorsModule.java]: Why not simply create a custom qualifier annotation and use it here?
 
```java
@Qualifier
@Documented
@Retention(RUNTIME)
public @interface OkHttpInterceptors {
}
```

**Comment by vanniktech** [comment]
@artem-zinnatullin looks good on my side. :+1:  I have only the one suggestion regarding the `@Named` in dagger. 

Since I haven't upgraded OkHttp in my projects to 3 yet, I haven't experienced this Robolectric issue.

File: app/src/debug/java/com/artemzin/qualitymatters/network/OkHttpInterceptorsModule.java (hunk 2 above)
Comment (artem-zinnatullin) [inline app/src/debug/java/com/artemzin/qualitymatters/network/OkHttpInterceptorsModule.java]: Well, it's 5 lines instead of one.. :) Idk, what benefits of qualifier do you see in this case?

Comment (vanniktech) [inline app/src/debug/java/com/artemzin/qualitymatters/network/OkHttpInterceptorsModule.java]: Well you don't need the constants anymore.
You can't misuse the constants since they don't exist.
It just reads a bit easier `@Provides @OkHttpNetworkInterceptors @Singleton @NonNull`.

Comment (artem-zinnatullin) [inline app/src/debug/java/com/artemzin/qualitymatters/network/OkHttpInterceptorsModule.java]: Ok, will use qualifiers!

On Sun, 7 Feb 2016, 23:19 Niklas Baudy <notifications@github.com> wrote:

> In
> app/src/debug/java/com/artemzin/qualitymatters/network/OkHttpInterceptorsModule.java
> <https://github.com/artem-zinnatullin/qualitymatters/pull/88#discussion_r52124050>
> :
>
> > + */
> > +@Module
> > +public class OkHttpInterceptorsModule {
> > +
> > +    // Provided as separate dependency for Developer Settings to be able to change HTTP log level at runtime.
> > +    @Provides @Singleton @NonNull
> > +    public HttpLoggingInterceptor provideHttpLoggingInterceptor() {
> > +        return new HttpLoggingInterceptor(message -> Timber.d(message));
> > +    }
> > +
> > +    @Provides @Named(LIST_OF_OK_HTTP_
>
> INTERCEPTORS) @Singleton @NonNull
> > +    public List<Interceptor> provideOkHttpInterceptors(@NonNull HttpLoggingInterceptor httpLoggingInterceptor) {
> > +        return singletonList(httpLoggingInterceptor);
> > +    }
> > +
> > +    @Provides @Named(LIST_OF_OK_HTTP_NETWORK_INTERCEPTORS) @Singleton @NonNull
>
> Well you don't need the constants anymore.
> You can't misuse the constants since they don't exist.
> It just reads a bit easier @Provides @OkHttpNetworkInterceptors
> @Singleton @NonNull.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/artem-zinnatullin/qualitymatters/pull/88/files#r52124050>
> .
>
-- 

@artem_zin

**Comment by artem-zinnatullin** [comment]
Found the issue with OkHttp https://github.com/square/okhttp/issues/2323, 3.0.1 doesn't have such problem.

// Another thing is that many classes in OkHttp are now final… uh

**Comment by artem-zinnatullin** [comment]
**ck in 3.1.0 `OkHttpClient` now back non-final, dependency trap :crying_cat_face:

**Comment by artem-zinnatullin** [comment]
I don't want to use PowerMock or other solutions, so, I'd wait for 3.1.1, should be released soon.

**Comment by artem-zinnatullin** [comment]
Unfortunately, OkHttp 3.1.1 didn't fix that, will continue monitoring the situation.

**Comment by vanniktech** [comment]
Hmm that's a bummer. Let's hope 3.1.2 will fix this.

**Comment by vanniktech** [comment]
@artem-zinnatullin I think 3.1.2 fixed the robolectric issue.

**Comment by artem-zinnatullin** [comment]
Yup, will try soon

On Thu, 11 Feb 2016, 07:28 Niklas Baudy <notifications@github.com> wrote:

> @artem-zinnatullin <https://github.com/artem-zinnatullin> I think 3.1.2
> fixed the robolectric issue.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/artem-zinnatullin/qualitymatters/pull/88#issuecomment-182836390>
> .
>
-- 

@artem_zin

File: app/src/main/java/com/artemzin/qualitymatters/api/ApiModule.java (hunk 1 above)
Comment (artem-zinnatullin) [inline app/src/main/java/com/artemzin/qualitymatters/api/ApiModule.java]: Created #92 for this, will go into separate PR

**Comment by artem-zinnatullin** [comment]
@vanniktech fixed comments!

**Comment by vanniktech** [comment]
Nice :+1:
